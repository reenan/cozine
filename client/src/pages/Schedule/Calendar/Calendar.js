import React, { Component } from 'react'
import './Calendar.scss';

import { Icon } from '../../shared';

import moment from 'moment';
import 'moment/locale/pt-br';
moment.locale('pt-BR');

//moment().year(value).month(value).daysInMonth();
//moment().year(value).month(value).startOf('month').day() => match weekDays

export default class Calendar extends Component {

  constructor(props) {
    super(props);

    this.state = {
      selectedMonth: 0,
      selectedYear: 2019
    }
  }

  static weekDays = ['DOM','SEG','TER','QUA','QUI','SEX','SAB'];
  static months = [
    'Janeiro', 
    'Fevereiro',
    'MarÃ§o',
    'Abril',
    'Maio',
    'Junho',
    'Julho',
    'Agosto',
    'Setembro',
    'Outubro',
    'Novembro',
    'Dezembro',
  ];

  prevMonth = () => {
    let { selectedMonth, selectedYear } = this.state;

    if (selectedMonth !== 0) {
      selectedMonth--;
    } else {
      selectedMonth = 11;
      selectedYear--;
    }

    this.setState({
      selectedMonth,
      selectedYear
    });
  }

  nextMonth = () => {
    let { selectedMonth, selectedYear } = this.state;

    if (selectedMonth !== 11) {
      selectedMonth++;
    } else {
      selectedMonth = 0;
      selectedYear++;
    }

    this.setState({
      selectedMonth,
      selectedYear
    });
  }

  render() {
    const { selectedMonth, selectedYear } = this.state;

    return (
      <div className='calendar'>
        <div className='header'>
          <p className='selected-year'>{selectedYear}</p>
          
          <Icon onClick={this.prevMonth} icon='arrow-left' />
          <p className='selected-month'>{Calendar.months[selectedMonth]}</p>
          <Icon onClick= {this.nextMonth}icon='arrow-right' />
        </div>

        <WeekDays weekDays={Calendar.weekDays} />
        <Month year={selectedYear} month={selectedMonth} />
      </div>
    )
  }
}

class Month extends Component {

  MONTH_SIZE = 6 * 7;

  addLastMonthPaddingDays = (list) => {
    const { year, month } = this.props;
    let currentMonthFirstDay = moment().year(year).month(month).startOf('month').day();
    let lastMonthDaysCount = moment().year(year).month(month).subtract(1, 'month').daysInMonth();
    let paddingValue = lastMonthDaysCount - currentMonthFirstDay;

    while (list.length < currentMonthFirstDay) {
      list.push(
        <li key={list.length} className='disabled'>
          <span>{++paddingValue}</span>
        </li>
      )
    }
  }

  addCurrentMonthDays = (list) => {
    const { year, month } = this.props;
    let monthDaysCount = moment().year(year).month(month).daysInMonth();

    for (let i = 1; i <= monthDaysCount; i++) {
      list.push(
        <li key={list.length}>
          <span>{i}</span>
        </li>
      )
    }
  }

  addNextMonthPaddingDays = (list) => {
    let paddingValue = 0;

    while (list.length < this.MONTH_SIZE) {
      list.push(
        <li key={list.length} className='disabled'>
          <span>{++paddingValue}</span>
        </li>
      )
    }
  }

  render() {
    let month = [];

    this.addLastMonthPaddingDays(month);
    this.addCurrentMonthDays(month);
    this.addNextMonthPaddingDays(month);

    return (
      <ul className='month'>
        {month}
      </ul>
    )

  }
}


const WeekDays = ({ weekDays }) => {

  return (
    <ul className='weekdays'>
      {
        weekDays.map((item, index) => {
          return (
            <li key={index}>
              <span>{item}</span>
            </li>
          )
        })
      }
    </ul>
  )
}
