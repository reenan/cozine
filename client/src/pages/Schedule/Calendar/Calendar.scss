@import '../../../resources/_variables.scss';
@import '../../../resources/_mixins.scss';

.calendar {
    display: inline-block;
    margin-top: 15px;

    .header {
        text-align: center;
        max-width: 300px;
        margin: auto;

        .selected-year {
            margin-bottom: 0;
            color: $main-color;
        }

        .selected-month, .icon {
            @include vertical-align(middle);
        }

        .selected-month {
            margin: 10px 0;
            font-size: 18px;
            width: 60%;
        }

        .icon {
            cursor: pointer;
            color: $main-color;
        }

    }

    ul.weekdays {
        > li {
            color: $main-color;
            font-size: 11px;
        }
    }

    ul.month {
        > li {
            @include transition(background-color 400ms);
            border-radius: 15%;
            position: relative;
            
            &:before {
                @include absolute-fill();
                content: '';
                border-radius: 15%;
            }

            &:not(.disabled):hover {
                background-color: rgba($main-color, 0.5);
            }

            &.disabled {
                color: $medium-gray;

                &:before {
                    background-color: $soft-gray;
                }
            }

            > span {
                position: relative;
            }
        }
    }

    ul.month,
    ul.weekdays {
        width: 100%;
        margin: auto;

        > li {
            @include box-sizing(border-box);
            @include vertical-align(middle);
    
            width: calc(100% / 7 - 20px);
            padding: 5px;
            margin: 10px;
            text-align: center;
            
        }

    }

    @media screen and (min-width: $tablet-width) {
        width: 50%;
    }
}