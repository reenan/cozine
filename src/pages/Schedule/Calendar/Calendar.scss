@import '../../../resources/_variables.scss';
@import '../../../resources/_mixins.scss';

.calendar {
  display: inline-block;
  color: $dark-text;
  width: 100%;
  text-align: center;

  &.popover {
    @include box-shadow(0 5px 10px 0, rgba(0, 0, 0, 0.1));
  }

  .header {
    max-width: 300px;
    margin: auto;

    .selected-year {
      margin-bottom: 0;
      color: $main-color;
    }

    .selected-month, .icon {
      @include vertical-align(middle);
    }

    .selected-month {
      margin: 10px 0;
      font-size: 18px;
      width: 60%;
    }

    .icon {
      cursor: pointer;
      color: $main-color;
    }

  }

  ul.weekdays {
    display: grid;
    grid-template-columns: repeat(7, 1fr);
    margin-top: $small-spacing;
    margin-bottom :$extra-small-spacing;

    > li {
      color: $main-color;
      font-size: 11px;
    }
  }

  ul.month {
    display: grid;
    grid-template-columns: repeat(7, 1fr);
    grid-auto-rows: 1fr;

    &:before {
      content: '';
      padding-bottom: 100%;
      grid-row: 1;
      grid-column: 1;
    }

    & > :first-child {
      grid-row: 1;
      grid-column: 1;
    }

    > li {
      position: relative;
      display: grid;
      align-content: center;

      &:before {
        @include transition(background-color 400ms);
        @include absolute-fill(5px);
        content: '';
        border-radius: 25%;
        z-index: -1;
      }

      &:not(.disabled):hover {
        &:before {
          background-color: rgba($main-color, 0.5);
        }
      }

      &.disabled {
        color: $medium-gray;

        &:before {
          background-color: $light-gray;
        }
      }
    }
  }

  @media screen and (min-width: $tablet-width) {
    .weekdays {
      margin-bottom: $small-spacing;
      span {
        font-size: 16px;
      }
    }

    .month span {
      font-size: 18px;
    }
  }
}