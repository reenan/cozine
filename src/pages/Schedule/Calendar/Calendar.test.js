import React from 'react';
import { render, cleanup } from 'react-testing-library';

import Calendar from './Calendar';

afterEach(cleanup)

it('renders correctly', () => {
  const { container } = render(<Calendar />);
  expect(container).toMatchSnapshot();
});