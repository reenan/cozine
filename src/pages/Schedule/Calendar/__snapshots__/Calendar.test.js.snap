// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    class="calendar"
  >
    <div
      class="header"
    >
      <p
        class="selected-year"
      >
        2019
      </p>
      <span
        class="icon arrow-left "
      />
      <p
        class="selected-month"
      >
        Janeiro
      </p>
      <span
        class="icon arrow-right "
      />
    </div>
    <ul
      class="weekdays"
    >
      <li>
        <span>
          DOM
        </span>
      </li>
      <li>
        <span>
          SEG
        </span>
      </li>
      <li>
        <span>
          TER
        </span>
      </li>
      <li>
        <span>
          QUA
        </span>
      </li>
      <li>
        <span>
          QUI
        </span>
      </li>
      <li>
        <span>
          SEX
        </span>
      </li>
      <li>
        <span>
          SAB
        </span>
      </li>
    </ul>
    <ul
      class="month"
    >
      <li
        class="disabled"
      >
        <span>
          30
        </span>
      </li>
      <li
        class="disabled"
      >
        <span>
          31
        </span>
      </li>
      <li>
        <span>
          1
        </span>
      </li>
      <li>
        <span>
          2
        </span>
      </li>
      <li>
        <span>
          3
        </span>
      </li>
      <li>
        <span>
          4
        </span>
      </li>
      <li>
        <span>
          5
        </span>
      </li>
      <li>
        <span>
          6
        </span>
      </li>
      <li>
        <span>
          7
        </span>
      </li>
      <li>
        <span>
          8
        </span>
      </li>
      <li>
        <span>
          9
        </span>
      </li>
      <li>
        <span>
          10
        </span>
      </li>
      <li>
        <span>
          11
        </span>
      </li>
      <li>
        <span>
          12
        </span>
      </li>
      <li>
        <span>
          13
        </span>
      </li>
      <li>
        <span>
          14
        </span>
      </li>
      <li>
        <span>
          15
        </span>
      </li>
      <li>
        <span>
          16
        </span>
      </li>
      <li>
        <span>
          17
        </span>
      </li>
      <li>
        <span>
          18
        </span>
      </li>
      <li>
        <span>
          19
        </span>
      </li>
      <li>
        <span>
          20
        </span>
      </li>
      <li>
        <span>
          21
        </span>
      </li>
      <li>
        <span>
          22
        </span>
      </li>
      <li>
        <span>
          23
        </span>
      </li>
      <li>
        <span>
          24
        </span>
      </li>
      <li>
        <span>
          25
        </span>
      </li>
      <li>
        <span>
          26
        </span>
      </li>
      <li>
        <span>
          27
        </span>
      </li>
      <li>
        <span>
          28
        </span>
      </li>
      <li>
        <span>
          29
        </span>
      </li>
      <li>
        <span>
          30
        </span>
      </li>
      <li>
        <span>
          31
        </span>
      </li>
      <li
        class="disabled"
      >
        <span>
          1
        </span>
      </li>
      <li
        class="disabled"
      >
        <span>
          2
        </span>
      </li>
      <li
        class="disabled"
      >
        <span>
          3
        </span>
      </li>
      <li
        class="disabled"
      >
        <span>
          4
        </span>
      </li>
      <li
        class="disabled"
      >
        <span>
          5
        </span>
      </li>
      <li
        class="disabled"
      >
        <span>
          6
        </span>
      </li>
      <li
        class="disabled"
      >
        <span>
          7
        </span>
      </li>
      <li
        class="disabled"
      >
        <span>
          8
        </span>
      </li>
      <li
        class="disabled"
      >
        <span>
          9
        </span>
      </li>
    </ul>
  </div>
</div>
`;
